File: 0454-4Sum-II/0454.py
class Solution:
    def fourSumCount(self, A, B, C, D):
        """
        :type A: List[int]
        :type B: List[int]
        :type C: List[int]
        :type D: List[int]
        :rtype: int
        """
        record = {}
        for i in C:
            for j in D:
                record[i + j] = record.get(i + j, 0) + 1

        result = 0
        for i in A:
            for j in B:
                if 0 - i - j in record:
                    result += record[0 - i - j]

        return result

if __name__ == "__main__":
    A = [ 1, 2]
    B = [-2,-1]
    C = [-1, 2]
    D = [ 0, 2]
    print(Solution().fourSumCount(A, B, C, D))

File: src/0454-4Sum-II/0454.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
using  namespace std;
static int x = []() {std::ios::sync_with_stdio(false); cin.tie(0); return 0; }();

class Solution 
{
public:
    int fourSumCount(vector<int>& A, vector<int>& B, vector<int>& C, vector<int>& D) 
    {
        unordered_map<int, int> record;
        for (auto& i : C)
        {
            for (auto& j : D)
            {
                ++record[i + j];
            }
        }
        int result = 0;
        for (auto& i : A)
        {
            for (auto& j : B)
            {
                if (record.count(- i - j))
                {
                    result += record[- i - j];
                }
            }
        }
        return result;
    }
};
int main()
{
    vector<int> A = {1, 2};
    vector<int> B = {-2, -1};
    vector<int> C = {-1, 2};
    vector<int> D = {0, 2};
    cout << Solution().fourSumCount(A, B, C, D) << endl;
    return 0;
}
